### ISSUE TITLE
Slower Build Performance after Upgrade from 9.0.7 to 9.1.0

### ISSUE THREAD

**Issue by sbradl** [angular/angular-cli#17385]
Since the update from Angular 9.0.7 to 9.1.0 our build times increased a lot. Here are some times:

v9.0.7 - prod build: 8:41
v9.1.0 - prod build: 13:09

These times are on a windows machine with node v12

On linux and mac the difference is much smaller - only a few seconds

**Comment by sbradl** [comment]
Are you sure it is an ngcc problem? The first line in my output is the normal ngc step and even this increased a lot

**Comment by alan-agius4** [comment]
The sync NGCC is run because of possible modules added in the compilation at a later stage after the first build. Example, when doing code changes, in a local library or adding a new NPM module.

That being said, if a module was already processed by the async NGCC should bail out pretty fast, it will just check for the build marker and exit. 

---

@blackholegalaxy 
> Fun fact: adding a new dependency in the project (dotenv) caused the entire ngcc compile to occur again even if it was the only changed dependency.

This is expected if you use `Yarn` as your package manager. This is because `Yarn` will remove the entire `node_modules` and re-install all dependencies when you add a new dependency. Which means that NGCC will need to re-run on all modules, which in this case NGCC will only run in sync mode.

**Comment by sbradl** [comment]
With #1 you mean the ngc step? This is the biggest problem for us because this also happens when we develop with --watch.

#4 could be. The windows system I used has an old ssd and my linux system has a new m2 ssd.

**Comment by alan-agius4** [comment]
> With `#1` you mean the ngc step? This is the biggest problem for us because this also happens when we develop with --watch.

Yes, me and @petebacondarwin discussed this in the past and @petebacondarwin had some ideas on how to speed this up when there is no work to be done. 

`#4` could be. The windows system I used has an old ssd and my linux system has a new m2 ssd.

From my experience in general IO operations are always slower on Windows.

**Comment by petebacondarwin** [comment]
So the actual ngcc processing should still only run once per package, so unless you modify your node_modules, the processing of these packages should not happen a second time.

Regarding 1. - this might well be the problem, since on Windows it is quite slow to  traverse the file-system checking for entry-points that "might" need processing. This is happening even if it turns out that they are all already processed.  That being said, I recently implemented an `__ngcc_entry_point_manifest__.json` file that should be written to the top of your `node_modules` directory that should avoid having to do this traversal... Can you check to see if one of these is appearing?

**Comment by petebacondarwin** [comment]
Oh, the other part of 1. is that once we have the list of entry-points, which can come from the manifest, we are still parsing the entry-points looking for dependencies in order to sort them before we determine whether they need to be processed or not. This is what @alan-agius4  is talking about here:

> Yes, me and @petebacondarwin discussed this in the past and @petebacondarwin had some ideas on how to speed this up when there is no work to be done. But in involved quite a decent amount of work at the time.

Perhaps now is the time to fix this?
